### ISSUE TITLE
Roadmap for chat components

### ISSUE THREAD

**Issue by ahuang11** [holoviz/panel#6291]
Highest priority:
1. Integrate a custom ChatInput https://github.com/holoviz/panel/issues/6115
  - 2-in-1 file upload + input text area (like GitHub)
  - keyboard shortcuts to send https://github.com/holoviz/panel/issues/6284
<img width="813" alt="image" src="https://github.com/holoviz/panel/assets/15331990/04c58cd7-e982-4a9d-8476-d4bcc2bb2d37"> 
2. Make it possible to customize styling of chat components https://github.com/holoviz/panel/issues/6235
3. [Migrate and integrate `Status` component](https://github.com/holoviz-topics/panel-chat-examples/blob/main/panel_chat_examples/components/status.py) that streams agents' intermediate steps or context from RAG (better naming TBD?)

Nice to haves:
1. Editing sent messages
2. ChatSuggestion buttons (click pre-defaults)
<img width="766" alt="image" src="https://github.com/holoviz/panel/assets/15331990/56e54f07-3e91-45cc-9ca4-2e63bd26961f">
5. Improved control over placeholder
6. Default `Help` user for instructions for user, and gets ignored upon `instance.serialize` https://github.com/holoviz/panel/issues/5984
7. Built-in export chat history button https://github.com/holoviz/panel/issues/5567
8. Custom callback exception handlers https://github.com/holoviz/panel/issues/6047

**Comment by MarcSkovMadsen** [comment]
Note: 2 in 1 Fileupload + input is something we lack in general not only for text. But for tables, images,video,Audio and maybe more. Gradio provides components like which makes model testing apps easy to create for their users.

**Comment by ahuang11** [comment]
> For non-chat use cases having a DragAndDropArea you could place a pane or (Tabulator) widget inside for display would be awesome.

I can't wrap my head around this. Can you elaborate?

> ChatSuggestions

Actually I'm not sure if it should be a component, or just a param of ChatFeed. What would this component add beyond being a button if it's not linked to anything?

**Comment by MarcSkovMadsen** [comment]
> > For non-chat use cases having a DragAndDropArea you could place a pane or (Tabulator) widget inside for display would be awesome.
> 
> I can't wrap my head around this. Can you elaborate?

Yes. `DragAndDropArea` is a layout that takes a *pane* or *widget* as input. It understand what  types of objects the given pane or widget can accept. When something valid is drag and dropped in its area its converted into some the pane or object can take as input.

```python
DragAndDropArea(Image("some_image.png")
```

If you drag an image into the above area it replaces `"some_image.png"`.

**Comment by ahuang11** [comment]
Sounds awesome!

**Comment by MarcSkovMadsen** [comment]
> > ChatSuggestions
> 
> Actually I'm not sure if it should be a component, or just a param of ChatFeed. What would this component add beyond being a button if it's not linked to anything?

Composability. It would be a lego brick that could be used with the `ChatFeed`. But also just with any *widget* that takes text as argument. Gradio often have apps with a `TextInput` the user can click some examples to quickly test things out.

**Comment by ahuang11** [comment]
Can you sketch out a param class for it? I still can't wrap my head around its benefit over using vanilla pn.widgets.Button.
